### PR TITLE
TEST/IODEMO: Add UCP AM support

### PR DESCRIPTION
**Pull request by brminich**
## What
Initial support for UCP AM in IODEMO

### CODE REVIEW

**Comment by brminich** [comment]
@evgeny-leksikov, can you please check the approach?

File: test/apps/iodemo/ucx_wrapper.h [hunk 1]
```diff
@@ -157,12 +164,18 @@ class UcxContext {
...
     std::deque<UcxConnection *>    _failed_conns;
+    bool                           _use_am;
 };
 
 
 class UcxConnection {
-public:
-    UcxConnection(UcxContext& context, uint32_t conn_id);
+    public:
+    struct ucx_am_desc_t {
```
Comment (evgeny-leksikov) [inline test/apps/iodemo/ucx_wrapper.h:173]: typedef

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -281,11 +284,8 @@ class DemoServer : public P2pDemoCommon {
     // sends an IO response when done
     class IoWriteResponseCallback : public UcxCallback {
     public:
-        IoWriteResponseCallback(size_t buffer_size,
-            MemoryPool<IoWriteResponseCallback>* pool) :
-            _server(NULL), _conn(NULL), _sn(0), _data_size(0), _chunk_cnt(0) {
-            _pool = pool;
-        }
+        IoWriteResponseCallback(size_t buffer_size) :
+            _server(NULL), _conn(NULL), _sn(0), _data_size(0), _chunk_cnt(0) {}
```
Comment (evgeny-leksikov) [inline test/apps/iodemo/io_demo.cc:288]: new line for `}`

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -336,13 +325,68 @@ class DemoServer : public P2pDemoCommon {
...
+            }
+            _pool->put(this);
+        }
+
+    protected:
+        MemoryPool<IoWriteTagResponseCallback>* _pool;
+    };
+
+    DemoServerTag(const options_t& test_opts) :
+        DemoServer(test_opts), _callback_pool(0) {}
```
Comment (evgeny-leksikov) [inline test/apps/iodemo/io_demo.cc:380]: new line for `}` as everywhere

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -768,17 +783,181 @@ class DemoClient : public P2pDemoCommon {
...
+
+        IoMessage *m = _io_msg_pool.get();
+        m->init(IO_READ, sn, data_size);
+
+        conn->send_am(m->buffer(), opts().iomsg_size, NULL, 0, m);
+        ++_num_sent;
+
+        return data_size;
+    }
+    virtual void handle_io_read_response(UcxConnection *conn,
```
Comment (evgeny-leksikov) [inline test/apps/iodemo/io_demo.cc:944]: pls add empty new line

**Review by evgeny-leksikov — COMMENTED** [review]
pls fix code style in other places as well

**Comment by brminich** [comment]
will be implemented on top of rebased IO_DEMO using another approach

**Pull request closed without merging**